### PR TITLE
observe imagePolicyConfig/internalRegistryHostname

### PR DESCRIPTION
**Pull request by sanchezl**
Rebase of pull request https://github.com/openshift/cluster-kube-apiserver-operator/pull/101.
TODO: add some unit tests.

### CODE REVIEW

**Comment by juanvallejo** [comment]
@sanchezl you might want to bump deps here to pick up [this fix to openshift/api](https://github.com/openshift/api/pull/123)

**Comment by bparees** [comment]
@sanchezl you're going to need to include a dependency bump.

**Comment by bparees** [comment]
@juanvallejo jinx

File: pkg/operator/observe_config.go [hunk 1]
```diff
@@ -257,13 +278,44 @@ func observeRestrictedCIDRs(kubeClient kubernetes.Interface, clientConfig *rest.
...
+// observeInternalRegistryHostname reads the internal registry hostname from the cluster configuration as provided by
+// the registry operator.
+func observeInternalRegistryHostname(listers Listers, currentConfig map[string]interface{}) (observedConfig map[string]interface{}, errs []error) {
+	observedConfig = map[string]interface{}{}
+	internalRegistryHostnamePath := []string{"imagePolicyConfig", "internalRegistryHostname"}
+	currentInternalRegistryHostname, _, _ := unstructured.NestedStringSlice(currentConfig, internalRegistryHostnamePath...)
+
+	configImage, err := listers.imageConfigLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		errs = append(errs, fmt.Errorf("image.config.openshift.io/cluster.kube-system: not found"))
```
Comment (bparees) [inline pkg/operator/observe_config.go]: cluster.kubesystem?  imageconfig is a cluster scoped resource.

Comment (bparees) [reply pkg/operator/observe_config.go]: also why are we returning an error here?  I thought we just wanted to log this.

Comment (bparees) [reply pkg/operator/observe_config.go]: also comments on each of these decisions would be helpful for the next person.. (why we are treating "not found" and "error retrieving" differently, for example)

Comment (sanchezl) [reply pkg/operator/observe_config.go]: @bparees , @deads2k : please review these scenarios

if the **image.config.openshift.io/cluster** resource does not exist, just log a message, do not set **imagePolicyConfig.internalRegistryHostName** and do not return an error.
```gherkin
# The internal registry is not expected to be setup by the openshift-installer.
Scenario: New cluster startup.
  Given image.config.openshift.io/cluster resource does not exist
  And imagePolicyConfig.internalRegistryHostName was not previously observed
  When an attempt is made to observe imagePolicyConfig.internalRegistryHostName
  Then imagePolicyConfig.internalRegistryHostName will not be set
  And no error will be returned
  And the Failing condition will not be set
```

but what if its not first startup and a previous **image.config.openshift.io/cluster** resource was deleted?
```gherkin
Scenario: Resource image.config.openshift.io/cluster deleted
  Given image.config.openshift.io/cluster resource does not exist
  And imagePolicyConfig.internalRegistryHostName was previously observed
  When an attempt is made to observe imagePolicyConfig.internalRegistryHostName
  Then imagePolicyConfig.internalRegistryHostName will not be set
  And no error will be returned
  And the Failing condition will not be set
```

if an error (other than NotFound) occurs getting the **image.config.openshift.io/cluster** resource, return the previously observed **imagePolicyConfig.internalRegistryHostName** if any.
```gherkin
Scenario: Error getting resource image.config.openshift.io/cluster (previously observed).
  Given imagePolicyConfig.internalRegistryHostName was previously observed
  When an attempt is made to observe imagePolicyConfig.internalRegistryHostName
  And an error occurs getting the image.config.openshift.io/cluster resource
  Then imagePolicyConfig.internalRegistryHostName will be set to the previously observed value
  And error will be returned
  And the Failing condition will be set

Scenario: Error getting resource image.config.openshift.io/cluster (not previously observed).
  Given imagePolicyConfig.internalRegistryHostName was not previously observed
  When an attempt is made to observe imagePolicyConfig.internalRegistryHostName
  And an error occurs getting the image.config.openshift.io/cluster resource
  Then imagePolicyConfig.internalRegistryHostName will not be set
  And error will be returned
  And the Failing condition will be set
```

If **status.internalRegistryHostName** is not set, then do not set **imagePolicyConfig.internalRegistryHostName**, even if there was a previous value for it.
```gherkin
Scenario: status.internalRegistryHostName not set in image.config.openshift.io/cluster
  Given image.config.openshift.io/cluster resource exists
  And status.internalRegistryHostName not set
  When an attempt is made to observe imagePolicyConfig.internalRegistryHostName
  Then imagePolicyConfig.internalRegistryHostName will not be set
  And no error will be returned
  And the Failing condition will not be set
```

Comment (bparees) [reply pkg/operator/observe_config.go]: > if the image.config.openshift.io/cluster resource does not exist, just log a message, do not set imagePolicyConfig.internalRegistryHostName and do not return an error.

sounds right.

> but what if its not first startup and a previous image.config.openshift.io/cluster resource was deleted?

don't care, still don't set it.  (this means the previous setting, if any, will be cleared from the apiserver config, right?).

> if an error (other than NotFound) occurs getting the image.config.openshift.io/cluster resource, return the previously observed imagePolicyConfig.internalRegistryHostName if any.

right.

> If status.internalRegistryHostName is not set, then do not set imagePolicyConfig.internalRegistryHostName, even if there was a previous value for it.

right.

**Comment by mfojtik** [comment]
@sanchezl can you split all changes in vendor/ into a separate commit? (call it `bump(*):`)

File: pkg/operator/observe_config.go (hunk 1 above)
Comment (sanchezl) [reply pkg/operator/observe_config.go]: Ok.

**Pull request merged**